packages/react/src/components/Table/TableMultiSortModal/TableMultiSortModal.jsx

```diff
@@ -0,0 +1,235 @@
+    multiSortSelectColumnThenByTitle: 'Then by',
+    multiSortDirectionLabel: 'Select a direction',
+    multiSortDirectionTitle: 'Sort order',
+    multiSortAddColumn: 'Add column',
+    multiSortRemoveColumn: 'Remove column',
+    multiSortAscending: 'Ascending',
+    multiSortDescending: 'Descending',
+  },
+};
+
+const TableMultiSortModal = ({ columns, sort, actions, showMultiSortModal, i18n }) => {
+  const {
+    onAddMultiSortColumn,
+    onRemoveMultiSortColumn,
+    onSaveMultiSortColumns,
+    onCancelMultiSortColumns,
+  } = actions;
+  const [selectedMultiSortColumns, setSelectedMultiSortColumns] = useState(sort);
+
+  useEffect(() => {
+    setSelectedMultiSortColumns(sort);
+  }, [sort]);
+
+  const sortDirections = useMemo(
+    () => [
+      {
+        id: 'ASC',
+        label: i18n.multiSortAscending,
+      },
+      {
+        id: 'DESC',
+        label: i18n.multiSortDescending,
+      },
+    ],
+    [i18n.multiSortAscending, i18n.multiSortDescending]
+  );
+
+  const handleSelectMultiSortColumn = useCallback(
+    (index) => ({ selectedItem }) => {
+      setSelectedMultiSortColumns((prev) => {
+        const currentItem = prev[index];
+        return Object.assign([], prev, {
+          [index]: {
+            ...currentItem,
+            columnId: selectedItem.id,
+          },
+        });
+      });
+    },
+    []
+  );
+  const handleSelectMultiSortColumnDirection = useCallback(
+    (index) => ({ selectedItem }) => {
+      setSelectedMultiSortColumns((prev) => {
+        const currentItem = prev[index];
+        return Object.assign([], prev, {
+          [index]: {
+            ...currentItem,
+            direction: selectedItem.id,
+          },
+        });
+      });
+    },
+    []
+  );
+
+  const multiSortColumns = useMemo(() => {
+    return columns
+      .filter((col) => {
+        const columnAlreadySorted =
+          selectedMultiSortColumns.find((sortCol) => sortCol.columnId === col.id) !== undefined;
+
+        return col.isSortable && !columnAlreadySorted;
+      })
+      .map((col) => ({ id: col.id, name: col.name }));
+  }, [columns, selectedMultiSortColumns]);
+
+  const handleAddMultiSortColumn = useCallback(
+    (index) => () => {
+      setSelectedMultiSortColumns((prev) => {
+        const clone = [...prev];
+        clone.splice(index + 1, 0, {
+          columnId: '',
+          direction: 'ASC',
+        });
+        return clone;
+      });
+
+      onAddMultiSortColumn(index);
+    },
+    [onAddMultiSortColumn]
+  );
+  const handleRemoveMultiSortColumn = useCallback(
+    (index) => () => {
+      setSelectedMultiSortColumns((prev) => {
+        const clone = [...prev];
+        clone.splice(index, 1);
+        return clone;
+      });
+
+      onRemoveMultiSortColumn(index);
+    },
+    [onRemoveMultiSortColumn]
+  );
+
+  const handleSaveMultiSortColumns = useCallback(() => {
+    onSaveMultiSortColumns(selectedMultiSortColumns);
```

Comment:
Perhaps we need to check that the selectedMultiSortColumns actually contain columns here. Not sure what the UI should look like, possibly a validation message warning that columns are must be selected. The current code can emit `[{columnId: "select", direction: "ASC"}, {columnId: "", direction: "ASC"}]` which works for the stateful table bit might cause problems for apps dealing with their own states.

Reply:
Yeah, I thought of that. This is one of those weird instances where we're building a library, but managing the application state, too. @davidicus What do you think? Do we need error handling on this modal onSave and if so, I assume that'll need to be exposed to consumers as well? Or, just expose an onSave callback and let them deal with it?

Reply:
Another option is to force the dropdown to have a preselected column, so there is no room for empty selections.

Reply:
default to first available column and direction